sql/core/src/main/scala/org/apache/spark/sql/execution/SparkStrategies.scala

```diff
@@ -259,6 +259,31 @@ abstract class SparkStrategies extends QueryPlanner[SparkPlan] {
           if (canBuildLeft(joinType)) BuildLeft else BuildRight
         }
 
+        /**
+         * See. [SPARK-32290]
+         * Not in Subquery will almost certainly be planned as a Broadcast Nested Loop join,
+         * which is very time consuming because it's an O(M*N) calculation.
+         * But if it's a single column NotInSubquery, and buildSide data is small enough,
+         * O(M*N) calculation could be optimized into O(M) using hash lookup instead of loop lookup.
+         */
+        def createBroadcastNullAwareHashJoin() = {
+          if (conf.notInSubqueryHashJoinEnabled &&
+            joinType == LeftAnti &&
+            canBroadcastBySize(right, conf) &&
+            right.output.length == 1) {
+            val (matched, _, _) =
+              joins.NotInSubqueryConditionPattern.singleColumnPatternMatch(condition)
```

Comment:
Is it typical for the conditions to be in the exec node ? I think not. 

If I understand the convention in the rest of the code: The conditions for creating the nodes are not typically included inside of the nodes themselves. The general idea is that the nodes just "do the work unconditionally", and the optimizer worries about exactly when/how to instantiate them.

Reply:
agree with you that ExecNode should unconditionally do execution jobs. I will update in next patch.